src/main/java/net/consensys/orion/http/handler/privacy/AddToPrivacyGroupHandler.java

```diff
@@ -0,0 +1,187 @@
+import net.consensys.orion.enclave.Enclave;
+import net.consensys.orion.enclave.PrivacyGroupPayload;
+import net.consensys.orion.enclave.QueryPrivacyGroupPayload;
+import net.consensys.orion.exception.OrionErrorCode;
+import net.consensys.orion.exception.OrionException;
+import net.consensys.orion.http.handler.set.SetPrivacyGroupRequest;
+import net.consensys.orion.network.ConcurrentNetworkNodes;
+import net.consensys.orion.network.NodeHttpClientBuilder;
+import net.consensys.orion.storage.Storage;
+import net.consensys.orion.utils.Serializer;
+
+import java.util.Arrays;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Stream;
+
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.ext.web.RoutingContext;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Add a member to a privacy group
+ */
+public class AddToPrivacyGroupHandler extends PrivacyGroupBaseHandler implements Handler<RoutingContext> {
+
+  private static final Logger log = LogManager.getLogger();
+
+  private final Storage<PrivacyGroupPayload> privacyGroupStorage;
+  private final Storage<QueryPrivacyGroupPayload> queryPrivacyGroupStorage;
+  private final Enclave enclave;
+
+
+  public AddToPrivacyGroupHandler(
+      final Storage<PrivacyGroupPayload> privacyGroupStorage,
+      final Storage<QueryPrivacyGroupPayload> queryPrivacyGroupStorage,
+      final ConcurrentNetworkNodes networkNodes,
+      final Enclave enclave,
+      final Vertx vertx,
+      final Config config) {
+    super(networkNodes, NodeHttpClientBuilder.build(vertx, config, 1500));
+    this.privacyGroupStorage = privacyGroupStorage;
+    this.queryPrivacyGroupStorage = queryPrivacyGroupStorage;
+    this.enclave = enclave;
+  }
+
+  @Override
+  public void handle(final RoutingContext routingContext) {
+    final byte[] request = routingContext.getBody().getBytes();
+    final ModifyPrivacyGroupRequest modifyPrivacyGroupRequest =
+        Serializer.deserialize(JSON, ModifyPrivacyGroupRequest.class, request);
+
+    final AtomicReference<PrivacyGroupPayload> combinedPrivacyGroup = new AtomicReference<>();
+    privacyGroupStorage.get(modifyPrivacyGroupRequest.privacyGroupId()).thenAccept((result) -> {
+      if (result.isEmpty() || result.get().state() == PrivacyGroupPayload.State.DELETED) {
+        routingContext
+            .fail(new OrionException(OrionErrorCode.ENCLAVE_PRIVACY_GROUP_MISSING, "privacy group not found"));
+        return;
+      }
+
+      final PrivacyGroupPayload oldPrivacyGroupPayload = result.get();
+      if (!Arrays.asList(oldPrivacyGroupPayload.addresses()).contains(modifyPrivacyGroupRequest.from())) {
+        routingContext
+            .fail(new OrionException(OrionErrorCode.ENCLAVE_PRIVACY_GROUP_MISSING, "privacy group not found"));
```

Comment:
Assuming that we don't want to acknowledge that the group exists to a potential attacker, yes.

Reply:
That is probably true - do we want to have a log message here indicating what the real failure was? Otherwise I'm going to pull out my hair when running tests and diagnosing client problems.